core/network/src/sync.rs

```diff
@@ -54,9 +54,15 @@ struct PeerSync<B: BlockT> {
 	pub recently_announced: VecDeque<B::Hash>,
 }
 
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum AncestorSearchState<B: BlockT> {
+	LinearSearch(NumberFor<B>),
+	BinarySearch(NumberFor<B>, NumberFor<B>),
+}
+
 #[derive(Copy, Clone, Eq, PartialEq, Debug)]
 enum PeerSyncState<B: BlockT> {
-	AncestorSearch(NumberFor<B>),
+	AncestorSearch(NumberFor<B>, AncestorSearchState<B>),
```

Comment:
When the state is `BinarySearch` the first number in this tuple is not actually used?
In that case it would be better to move the starting number under `LinearSearch` enum value and just use the enum here instead of the tuple.

Reply:
It's used, for instance here https://github.com/paritytech/substrate/blob/754163db2490357f881396f32fb77ab85d5835a4/core/network/src/sync.rs#L486. I'm also not sure what is best, but I'm keeping as it is, otherwise I'll need to add more arguments and code to `handle_ancestor_search_state`.